File: packages/cli/src/merge.ts
import diff from 'fast-diff'
import type * as Y from 'yjs'

interface Edit {
  start: number
  end: number
  insert: string
}

/** Edits turning `from` into `to`, in `from` offsets, in ascending order. */
function editsBetween(from: string, to: string): Edit[] {
  const edits: Edit[] = []
  let pos = 0
  for (const [op, chunk] of diff(from, to)) {
    if (op === diff.EQUAL) {
      pos += chunk.length
    } else if (op === diff.DELETE) {
      const last = edits.at(-1)
      if (last && last.end === pos) last.end += chunk.length
      else edits.push({ start: pos, end: pos + chunk.length, insert: '' })
      pos += chunk.length
    } else {
      const last = edits.at(-1)
      if (last && last.end === pos) last.insert += chunk
      else edits.push({ start: pos, end: pos, insert: chunk })
    }
  }
  return edits
}

/** Returns a function mapping offsets in `base` to offsets in `current`. */
function offsetMapper(base: string, current: string): (offset: number) => number {
  const ops = diff(base, current)
  return (offset) => {
    let b = 0
    let c = 0
    for (const [op, chunk] of ops) {
      const len = chunk.length
      if (op === diff.EQUAL) {
        if (offset <= b + len) return c + (offset - b)
        b += len
        c += len
      } else if (op === diff.DELETE) {
        if (offset <= b + len) return c
        b += len
      } else {
        c += len
      }
    }
    return c + (offset - b)
  }
}

/**
 * Applies the change from `base` to `next` (an edit made to the file outside ima)
 * onto `text`, which may meanwhile have diverged from `base` through remote edits.
 * Remote edits are kept unless the external edit replaced the same region.
 */
export function mergeExternalEdit(text: Y.Text, base: string, next: string, origin: unknown): void {
  const current = text.toString()
  const edits = editsBetween(base, next)
  if (edits.length === 0) return
  const map = offsetMapper(base, current)
  const mapped = edits.map((e) => ({
    start: map(e.start),
    end: Math.max(map(e.start), map(e.end)),
    insert: e.insert,
  }))
  const doc = text.doc
  const apply = () => {
    for (let i = mapped.length - 1; i >= 0; i--) {
      const e = mapped[i] as Edit
      if (e.end > e.start) text.delete(e.start, e.end - e.start)
      if (e.insert) text.insert(e.start, e.insert)
    }
  }
  if (doc) doc.transact(apply, origin)
  else apply()
}

File: internal/merge/merge.go
// Package merge applies edits made to the file outside ima onto the shared text.
package merge

import (
	"unicode/utf16"
	"unicode/utf8"

	"github.com/reearth/ygo/crdt"
	"github.com/sergi/go-diff/diffmatchpatch"
)

// edit replaces [start, end) with insert. Offsets are in runes.
type edit struct {
	start, end int
	insert     string
}

func diffRunes(from, to []rune) []diffmatchpatch.Diff {
	return diffmatchpatch.New().DiffMainRunes(from, to, false)
}

// editsBetween returns the edits turning from into to, in from offsets, in ascending order.
func editsBetween(from, to []rune) []edit {
	var edits []edit
	pos := 0
	for _, d := range diffRunes(from, to) {
		n := utf8.RuneCountInString(d.Text)
		switch d.Type {
		case diffmatchpatch.DiffEqual:
			pos += n
		case diffmatchpatch.DiffDelete:
			if last := len(edits) - 1; last >= 0 && edits[last].end == pos {
				edits[last].end += n
			} else {
				edits = append(edits, edit{start: pos, end: pos + n})
			}
			pos += n
		case diffmatchpatch.DiffInsert:
			if last := len(edits) - 1; last >= 0 && edits[last].end == pos {
				edits[last].insert += d.Text
			} else {
				edits = append(edits, edit{start: pos, end: pos, insert: d.Text})
			}
		}
	}
	return edits
}

// offsetMapper returns a function mapping rune offsets in base to rune offsets in current.
func offsetMapper(base, current []rune) func(int) int {
	diffs := diffRunes(base, current)
	return func(offset int) int {
		b, c := 0, 0
		for _, d := range diffs {
			n := utf8.RuneCountInString(d.Text)
			switch d.Type {
			case diffmatchpatch.DiffEqual:
				if offset <= b+n {
					return c + (offset - b)
				}
				b += n
				c += n
			case diffmatchpatch.DiffDelete:
				if offset <= b+n {
					return c
				}
				b += n
			case diffmatchpatch.DiffInsert:
				c += n
			}
		}
		return c + (offset - b)
	}
}

// utf16Offsets maps each rune offset in rs (0..len(rs)) to a UTF-16 offset,
// the unit Y.Text indexes by.
func utf16Offsets(rs []rune) []int {
	out := make([]int, len(rs)+1)
	for i, r := range rs {
		out[i+1] = out[i] + utf16.RuneLen(r)
	}
	return out
}

// ExternalEdit applies the change from base to next (an edit made to the file
// outside ima) onto text, which may meanwhile have diverged from base through
// remote edits. Remote edits are kept unless the external edit replaced the same
// region. The caller must keep text from changing concurrently.
func ExternalEdit(doc *crdt.Doc, text *crdt.YText, base, next string, origin any) {
	baseRunes := []rune(base)
	edits := editsBetween(baseRunes, []rune(next))
	if len(edits) == 0 {
		return
	}
	current := []rune(text.ToString())
	toCurrent := offsetMapper(baseRunes, current)
	toUTF16 := utf16Offsets(current)
	doc.Transact(func(txn *crdt.Transaction) {
		for i := len(edits) - 1; i >= 0; i-- {
			e := edits[i]
			start := toCurrent(e.start)
			end := max(start, toCurrent(e.end))
			if end > start {
				text.Delete(txn, toUTF16[start], toUTF16[end]-toUTF16[start])
			}
			if e.insert != "" {
				text.Insert(txn, toUTF16[start], e.insert, nil)
			}
		}
	}, origin)
}
